buffling/privacy.py

```diff
@@ -0,0 +1,265 @@
+    y_x: An integer, the # of 1s in bloom filter after "shuffling+blipping".
+    p: An real number in [0,1], bit flipping probability.
+    n_bit: Integer, the length of each bloom filter.
+    depth: Positive integer, Approximating depth. 
+      Will be Passed to evaluate_gauss_continued_fraction().
+
+  Returns:
+    A real number of the simulated epsilon.
+  """
+  q = 1 - p
+  if y_d + y_x > n_bit: 
+    y_d = n_bit - y_d
+    y_x = n_bit - y_x
+  
+  a = - y_d
+  b = - y_x
+  c = n_bit - y_x - y_d 
+  z = (q / p) ** 2
+  l = evaluate_gauss_continued_fraction(a, b, c, z, depth=depth)
+
+  return (n_bit - y_x) / np.maximum(n_bit - y_x - y_d, 1) * l
+
+
+def evaluate_eps_given_yd(y_d, p, n_bit, type="exact", y_x=None, depth=1000):
+  """Evaluate the privacy parameter (epsilon) of one BF.
+
+  Args: 
+    y_d: An integer, the observed # of 1s in bloom filter.
+    p: An real number in [0,1], bit flipping probability.
+    n_bit: Integer, the length of each bloom filter.
+    type: A character string, either "exact" or "asymp".
+    y_x: An integer, the # of 1s in bloom filter after "shuffling+blipping".
+    depth: Integer, Approximating depth. 
+      Will be Passed to evaluate_gauss_continued_fraction().
+
+  Returns:
+    A real number of the simulated epsilon.
+  """
+  ## process the inputs
+  q = 1 - p
+  if y_x == None:
+    y_x = np.random.binomial(y_d, q) + np.random.binomial(n_bit - y_d, p)
+    if y_d + y_x > n_bit: 
+      y_d = n_bit - y_d
+      y_x = n_bit - y_x
+
+  if (type == "asymp"): ## normal approximation
+    b = 2 * np.log(q / p)
+    l = -b * y_d / n_bit + b ** 2 / 2 * y_d * (n_bit - y_d) * (2 * y_x - n_bit) / n_bit ** 3
+    l = np.exp(l)
+  elif (type == "exact"): ## exact form
+    l = ratio_of_hypergeometric_mgf(y_d=y_d, y_x=y_x, p=p, n_bit=n_bit, depth=depth) 
+  
+  ## Pr(A(D')=X) / Pr(A(D)=X)
+  privacy_loss = p / q / l ## version 2
+
+  return np.log(privacy_loss)
+
+
+def find_yx_given_yd_Is_1(p, n_bit, d):
+  '''Find the 1-d quantile of y_x given y_d = 1.
+  
+  Args: 
+    p: An real number in [0,1], bit flipping probability.
+    n_bit: Integer, the length of each bloom filter.
+    d: A real number, the approximation parameter.
+  
+  Returns: 
+    An integer, the # of 1s in bloom filter after "shuffling+blipping".
+  '''
+  q = 1 - p
+  res = np.zeros(2)
+  ## pmf of y_x
+  v = stats.binom(n_bit - 1, p)
+  ## find d/2 quantile
+  lower = 0
+  upper = n_bit
+  while lower + 1 < upper: 
+    y_x = (lower + upper) // 2
+    y_x_cdf = p * v.cdf(y_x) + q * v.cdf(y_x - 1) 
+    if y_x_cdf > d / 2:
+      upper = y_x
+    else: 
+      lower = y_x
+  res[0] = y_x
+
+  ## find 1-d/2 quantile
+  lower = 0
+  upper = n_bit
+  while lower + 1 < upper: 
+    y_x = (lower + upper) // 2
+    y_x_cdf = p * v.cdf(y_x) + q * v.cdf(y_x - 1) 
+    if y_x_cdf > 1 - d / 2:
+      upper = y_x
+    else: 
+      lower = y_x
+  res[1] = y_x
+  
+  return res
+
+
+def find_p_given_yd_is_1(e, n_bit, d=1e-4):
```

Comment:
Something like find_flip_prob_given_observed_count?

Reply:
Thanks, I will change the function names.